util/progress/progressui/printer.go

```diff
@@ -338,3 +353,52 @@ func limitString(s string, l int) string {
 	}
 	return s
 }
+
+// textPrinter prints the progress to a raw [io.Writer].
+type textPrinter interface {
+	// Print will print a message for the given build step at the index.
+	//
+	// If no index is associated with the message, a negative number will
+	// be used.
+	Print(w io.Writer, index int, msg string) error
+}
+
+// rawPrinter prints the text directly to the [io.Writer] with minor formatting.
+type rawPrinter struct{}
+
+func (*rawPrinter) Print(w io.Writer, index int, msg string) error {
+	if index < 0 {
+		_, err := fmt.Fprintf(w, "%s\n", msg)
+		return err
+	}
+	_, err := fmt.Fprintf(w, "#%d %s\n", index, msg)
+	return err
+}
+
+// jsonPrinter prints the text in JSON format.
+type jsonPrinter struct{}
+
+func (*jsonPrinter) Print(w io.Writer, index int, msg string) error {
```

Comment:
JSON printing should keep the original formatting. Instead of converting to `index-msg` pairs and printing these as JSON it should print the original `Vertex/VertexLog` etc objects as soon as they appear.

Another logic would be to try to avoid partial events and only print vertex once it has completed. This would probably be easier to analyze from caller side.

Eg.
```
{
  "digest": "",
  "name": "",
  "intervals": [
    {start:, complete:, cached:, error:}
  ],
  "logs": [
    {timestamp:, msg}
  ]
}
```

Reply:
Looking at my previous example, it is hard to know when last interval has completed and that would basically mean that all steps would need to be completed before logs can appear that is not   ideal in some cases.

Instead we could do something similar but only once per interval. So every time there is a completed event for vertex a new record is written out (with all the logs for that vertex). If step has multiple run intervals, there can be multiple objects with same digest.

We could also just do the raw event objects instead but maybe in that case we should call this `rawjson`.

Reply:
I mocked together an output format that I think will work well. As an example:

```
{"digest":"sha256:0a1c2e924062567ebce3f0e59cf5a8215b0ca468feb1cfd839572e11b6926659","name":"[internal] load .dockerignore","statuses":[{"ID":"transferring context:","Vertex":"sha256:0a1c2e924062567ebce3f0e59cf5a8215b0ca468feb1cfd839572e11b6926659","Name":"transferring","Tot
al":0,"Current":93,"Timestamp":"2023-08-10T17:23:44.811591403Z","Started":"2023-08-10T17:23:44.806562361Z","Completed":"2023-08-10T17:23:44.811591153Z"}],"done":true}                                                                                                            {"digest":"sha256:2eaf60b1da55757d511ec1247ba9e556f67ca9416ec8e78c3de743aa324cc01d","name":"[git 1/2] FROM docker.io/library/alpine:3.18@sha256:7144f7bab3d4c2648d7e59409f15ec52a18006a128c733fcff20d3a4a54ba44a","statuses":[{"ID":"resolve docker.io/library/alpine:3.18@sha256:
7144f7bab3d4c2648d7e59409f15ec52a18006a128c733fcff20d3a4a54ba44a","Vertex":"sha256:2eaf60b1da55757d511ec1247ba9e556f67ca9416ec8e78c3de743aa324cc01d","Name":"","Total":0,"Current":0,"Timestamp":"2023-08-10T17:23:45.088739986Z","Started":"2023-08-10T17:23:45.071948486Z","Completed":"2023-08-10T17:23:45.088739736Z"}],"done":true}                                                                                   {"digest":"sha256:4fdbd27f357e8ca142a8dfde9064febf498ca55d6ac098fe79444f605d732794","name":"[xx 1/1] FROM docker.io/tonistiigi/xx:1.2.1@s
ha256:8879a398dedf0aadaacfbd332b29ff2f84bc39ae6d4e9c0a1109db27ac5ba012","statuses":[{"ID":"resolve docker.io/tonistiigi/xx:1.2.1@sha256:8879a398dedf0aadaacfbd332b29ff2f84bc39ae6d4e9c0a1109db27ac5ba012","Vertex":"sha256:4fdbd27f357e8ca142a8dfde9064febf498ca55d6ac098fe79444f6
05d732794","Name":"","Total":0,"Current":0,"Timestamp":"2023-08-10T17:23:45.092603778Z","Started":"2023-08-10T17:23:45.072707403Z","Completed":"2023-08-10T17:23:45.092603361Z"}],"done":true}
```

The basic gist of it is that each JSON blob has a digest attached to it and then the rest are only sections which correspond to updates on the vertex. The intervals are presently omitted but they can likely be attached to the done attribute or included as their own attribute if needed. It should be feasibly possible to use the JSON event stream to update a third-party UI with the information provided.

The name is included but only the first time that a vertex is printed. One of cached, done, or error will be printed when the vertex finishes. Statuses and logs are included as part of the blob.

I think this output will probably be best. The biggest concern for JSON outputs are repeated information (such as printing the digest over and over again) and redundant information (such as the full vertex struct). I also chose to keep partial log lines in as I think it's always possible for a third-party to choose to buffer partial log lines, but it's not possible to retrieve partial lines if the formatter isn't even giving them to you. Since the primary purpose of this output is to enable third-party progress bars, I thought it would be more simple to keep it and it can always be removed later without changing the external API.

This is only a mockup so I don't have code that I feel comfortable pushing for review yet.